File: java/Weed3/src/noear/weed/DbContext.java
package noear.weed;

import noear.weed.ext.Act1;
import noear.weed.ext.Act1Ex;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;

/**
 * Created by noear on 14-6-12.
 * 数据库上下文
 */
public class DbContext {
    public Command lastCommand;
    public boolean allowMultiQueries;

    public DbContext(String schemaName,String url) {
        _schemaName = schemaName;
        _url = url;

        _fieldFormat = "";
    }

    //基于线程池配置（如："proxool."）
    //fieldFormat："`%`"
    public DbContext(String schemaName,String url,String fieldFormat) {
        _schemaName = schemaName;
        _url = url;

        _fieldFormat = fieldFormat;
    }

    //基于手动配置（无线程池）
    public DbContext(String schemaName,String url, String user,String password, String fieldFormat) {
        _schemaName = schemaName;
        _url = url;
        _user = user;
        _password = password;

        _fieldFormat = fieldFormat;
    }

    private String _url;
    private String _user;
    private String _password;

    private String _schemaName;

    private String _fieldFormat;

    public String field(String key){
        if(_fieldFormat == null || _fieldFormat.length()==0)
            return key;
        else
            return _fieldFormat.replace("%",key);
    }

    /*是否配置了schema*/
    public boolean hasSchema(){return _schemaName!=null;}

    /*获取schema*/
    public String getSchema(){
        return _schemaName;
    }

    protected Object pool;
    /*获取连接*/
    public  Connection getConnection() throws SQLException {
        if (_user == null)
            return DriverManager.getConnection(_url);
        else
            return DriverManager.getConnection(_url, _user, _password);
    }

    public DbQuery sql(String code, Object... args) {
        return new DbQuery(this).sql(new SQLBuilder().append(code, args));
    }

    public DbQuery sql(Act1<SQLBuilder> sqlBuilder) {
        SQLBuilder sql = new SQLBuilder();
        sqlBuilder.run(sql);
        return new DbQuery(this).sql(sql);
    }

    public DbQuery sql(SQLBuilder sqlBuilder) {
        return new DbQuery(this).sql(sqlBuilder);
    }

    /*获取process执行对象*/
    public DbProcedure call(String process) {
        if(process.indexOf(" ")>0)
            return new DbQueryProcedure(this).sql(process);
        else
            return new DbStoredProcedure(this).call(process);
    }


    /*获取一个表对象［用于操作插入也更新］*/
    public DbTableQuery table(String table) {
        return new DbTableQuery(this).table(table);
    }

    public DbTran tran(Act1Ex<DbTran,SQLException> handler) throws SQLException {
        return new DbTran(this).execute(handler);
    }

    public DbTran tran()
    {
        return new DbTran(this);
    }

}

File: .net standard2/Weed3/DbContext.cs
using System;
using System.Collections.Generic;
using System.Configuration;
using System.Data.Common;

namespace Noear.Weed {
    
    /**
     * Created by noear on 14-6-12.
     * 数据库上下文
     */
    public class DbContext {

        private string _fieldFormat;
        private String _url;
        private String _schemaName;
        private DbProviderFactory _provider;

        static DbProviderFactory provider(string providerString) {
            if (providerString == null)
                return null;
            
            //if (providerString.IndexOf(",") > 0)
                return (DbProviderFactory)Activator.CreateInstance(Type.GetType(providerString, true, true));
            //else
                //return DbProviderFactories.GetFactory(providerString);
        }

        public Command lastCommand { get; internal set; }
        public bool allowMultiQueries;

        /*
        public DbContext(String schemaName, string name) : this(schemaName, name, "") {

        }

        //fieldFormat："`%`"
        public DbContext(String schemaName, string name, string fieldFormat) {
            var set = ConfigurationManager.ConnectionStrings[name];
            var p = provider(set.ProviderName);
            doInit(schemaName, set.ConnectionString, p);

            _fieldFormat = fieldFormat;
        }*/

        public DbContext(String schemaName, string connectionString, DbProviderFactory provider) : this(schemaName, connectionString, "", provider) {

        }

        public DbContext(String schemaName, string connectionString, String fieldFormat, DbProviderFactory provider) {
            doInit(schemaName, connectionString, provider);
            _fieldFormat = fieldFormat;
        }
        
        protected virtual void doInit(String schemaName, string connectionString, DbProviderFactory provider) {
            _provider = provider;
            _schemaName = schemaName;
            _url = connectionString;
        }

        public String field(String key) {
            if (string.IsNullOrEmpty(_fieldFormat))
                return key;
            else
                return _fieldFormat.Replace("%", key);
        }

        /*是否配置了schema*/
        public bool hasSchema() { return _schemaName != null; }

        /*获取schema*/
        public String getSchema() {
            return _schemaName;
        }

        /*获取连接*/
        public DbConnection getConnection() {
            DbConnection conn = _provider.CreateConnection();

            conn.ConnectionString = _url;
            return conn;
        }

        public DbQuery sql(String code, params object[] args) {
            return new DbQuery(this).sql(new SQLBuilder().append(code, args));
        }

        public DbQuery sql(SQLBuilder sqlBuilder) {
            return new DbQuery(this).sql(sqlBuilder);
        }

        /*获取process执行对象*/
        public DbStoredProcedure call(String process) {
            return new DbStoredProcedure(this).call(process);
        }

        /*获取一个表对象［用于操作插入也更新］*/
        public DbTableQuery table(String table) {
            return new DbTableQuery(this).table(table);
        }

        public DbTran tran(Action<DbTran> handler) {
            return new DbTran(this).execute(handler);
        }

        public DbTran tran() {
            return new DbTran(this);
        }
        
    }
}
